### PR TITLE
Sequential v2

### PR DESCRIPTION
**Pull request by rhensen**
This PR reaches the first version of many of the v1 soil and crop objectives: 

Management practices:

- Crop types - corn (grain and silage), alfalfa, grass (tall fescue_, soybeans, winter wheat, winter rye, triticale
- Cover cropping
- No-till vs. tillage vs. conservation tillage
- Manure and fertilizer application - surface, incorporation, injections

Outcomes: 

- N2O, NH3, and CO2 Emissions
- N & P Leaching and Runoff
- Water use
- Soil C
- Crop yields and compositions

In addition, unit tests are provided for some of the recent changes. We hope to add more of these in the future.

### CODE REVIEW

File: RUFAS/routines/field/field_management/field_management.py [hunk 1]
```diff
@@ -101,8 +101,8 @@ def __init__(self, app_data, time):
             "tillage": {
                 'year': -1,
                 'day': -1,
-                'percent_incorporated': 0.8,
-                'percent_mixed': 0.6,
+                'percent_manure_incorporated': 0.8, #need to check this one out and potentially change to nutrient
```
Comment (rhensen) [inline RUFAS/routines/field/field_management/field_management.py]: percent_nutrient_incorporated
percent_nutrient_mixed

File: RUFAS/output_handler/reports/field_report.py [hunk 2]
```diff
@@ -324,7 +325,169 @@ def __init__(self, data, field_name):
                                         'fr_N': ['crop_type.fr_N', '', []],
                                         'precipitation': ['weather.rainfall[time.year - 1][time.day - 1]', 'mm', []],
                                         'T_d': ['soil.T_d', '', []],
-                                        }
+                                        'AG_met_layer0': ['soil.soil_layers[0].AG_met', '%', []],
```
Comment (PooyaHekmati) [inline RUFAS/output_handler/reports/field_report.py:329]: There are two levels of repetition here, both can be mitigated using zip() and lambda functions with some loops. Also, to add clarity, this big dict structure initialization should be wrapped in a function with an adequately intuitive name.

File: RUFAS/output_handler/reports/field_report.py [hunk 3]
```diff
@@ -324,7 +325,169 @@ def __init__(self, data, field_name):
...
+                                            'soil.soil_layers[3].C_passive', '%', []],
+
+                                        'C_passive_decomp_layer0': [
+                                            'soil.soil_layers[0].C_passive_decomp', '%', []],
+                                        'C_passive_decomp_layer1': [
+                                            'soil.soil_layers[1].C_passive_decomp', '%', []],
+                                        'C_passive_decomp_layer2': [
+                                            'soil.soil_layers[2].C_passive_decomp', '%', []],
+                                        'C_passive_decomp_layer3': [
+                                            'soil.soil_layers[3].C_passive_decomp', '%', []]                                        }
```
Comment (PooyaHekmati) [inline RUFAS/output_handler/reports/field_report.py]: misplaced curly brace

File: RUFAS/routines/field/crop/crop.py [hunk 4]
```diff
@@ -6,10 +6,12 @@
            Andy Achenreiner, achenreiner@wisc.edu
            William Donovan, wmdonovan@wisc.edu
            Jacob Johnson, jacob8399@gmail.com
+           Michael Tang, mstang2@wisc.edu
 """
 
 from math import acos, asin, sin, tan, pi
-from .crop_types import base_crop, alfalfa, corn, soybean, tall_fescue, spring_barley, potato, sugar_beet, spring_wheat
+from .crop_types import base_crop, alfalfa, corn, soybean, tall_fescue, spring_barley, potato, sugar_beet, spring_wheat,\
```
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop.py]: Line too long, please see https://peps.python.org/pep-0008/#maximum-line-length

File: RUFAS/routines/field/crop/crop.py [hunk 5]
```diff
@@ -31,12 +33,31 @@ def daily_crop_routine(soil, crop, field_management, weather, time):
     # Current crop is set at the beginning of the year in annual_crop_routine
     crop_type = crop.current_crop
 
+    # If there is no crop in rotation, current crop will be named
```
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop.py:36]: there are no if conditions next to this comment, how is this decision being made?

File: RUFAS/routines/field/crop/crop.py [hunk 6]
```diff
@@ -31,12 +33,31 @@ def daily_crop_routine(soil, crop, field_management, weather, time):
...
     daily_reset(crop_type, soil)
-
-    # If there is no crop in rotation this year, current crop will be named
     # 'null'. The routine is skipped in this case
     if crop_type.crop_name != 'null':
 
+        #if the crop was killed the previous day, set the new crop
```
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop.py]: remove comment by something like this:
crop_was_killed_yesterday = crop_type.killed
if crop_was_killed_yesterday:

File: RUFAS/routines/field/crop/crop.py [hunk 7]
```diff
@@ -31,12 +33,31 @@ def daily_crop_routine(soil, crop, field_management, weather, time):
...
+        if crop_type.killed:
+            # set the next crop to grow in a double cropping rotation
+            if crop_type.harvest_day > crop_type.planting_day:
```
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop.py:45]: can we wrap this in a function and call it set_next_crop or something like that?

File: RUFAS/routines/field/crop/crop.py [hunk 8]
```diff
@@ -31,12 +33,31 @@ def daily_crop_routine(soil, crop, field_management, weather, time):
...
+                if crop_type.planting_order == '1st':
+                    if crop.current_crop_year[1].crop_name != 'null':
+                        crop_type.killed = False
+                        crop.current_crop = crop.current_crop_year[1]
```
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop.py]: unclear and unexplained hardcoded value, what is the significance of 1? what happens if we change it to 0?

File: RUFAS/routines/field/crop/crop.py [hunk 9]
```diff
@@ -31,12 +33,31 @@ def daily_crop_routine(soil, crop, field_management, weather, time):
...
+                        crop.current_crop.kill_year = is_kill_year(crop, time)
+
+            # current crop year will be this years current crop while the current crop
```
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop.py]: vague and unclear comment

File: RUFAS/routines/field/crop/crop.py [hunk 10]
```diff
@@ -46,10 +67,14 @@ def daily_crop_routine(soil, crop, field_management, weather, time):
         crop_type.HI_actual = 0
         crop_type.bio_BG = 0
         soil.residue_harvest = 0
-        soil.soil_layers[0].tillage_percent = 0
+        soil.soil_layers[0].fr_tillage = 0
+
+        # If the crop is not planted yet, determine whether it is planted today
```
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop.py]: unnecessary comment

File: RUFAS/routines/field/crop/crop.py [hunk 11]
```diff
@@ -46,10 +67,14 @@ def daily_crop_routine(soil, crop, field_management, weather, time):
...
+        if not crop_type.planted:
+            calculate_start(soil, crop, field_management, weather, time)
 
-        # While the crop is planted:
-        if crop_type.planted and not crop_type.harvested:
+        # Once the crop is planted:
```
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop.py]: unnecessary and probably misleading comment. if crop_type.planted is true, how do we know it has been just planted? Moreover, can a type of crop really be planted? A crop can be planted, the type is describing it. A type is not plantable, a crop is.

File: RUFAS/routines/field/crop/crop.py [hunk 12]
```diff
@@ -46,10 +67,14 @@ def daily_crop_routine(soil, crop, field_management, weather, time):
...
+        else:
 
             # If the crop is growing, run the routines
```
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop.py]: unnecessary comment

**Review by PooyaHekmati — COMMENTED** [review]
I made these comments as a partial review to unblock this PR. Will get back to it and add more soon.

File: RUFAS/output_handler/reports/field_report.py [hunk 13]
```diff
@@ -304,7 +305,7 @@ def __init__(self, data, field_name):
                                          'fert_P_leached': ['soil.fert_P_leached_annual', 'kg', 0],
                                          'M_leach': ['soil.M_leach_annual', 'kg', 0],
                                          'DRP_drainage': ['soil.DRP_drainage_annual', 'kg', 0],
-                                         'M_DRP_runoff': ['soil.M_DRP_runoff_annual', 'kg', 0],
+                                                 'M_DRP_runoff': ['soil.M_DRP_runoff_annual', 'kg', 0],
```
Comment (HectorMM3) [inline RUFAS/output_handler/reports/field_report.py:309]: Why is M_DRP_runoff tabbed over? Does it need to be?

File: RUFAS/routines/field/crop/crop.py (hunk 7 above)
Comment (HectorMM3) [reply RUFAS/routines/field/crop/crop.py:45]: Pooya, I think we want to leave it as is for now because we can going to make version 3 more complex at that time we will attempt to put it in a function.

File: RUFAS/routines/field/crop/crop.py [hunk 14]
```diff
@@ -110,7 +134,7 @@ def daily_reset(crop_type, soil):
     crop_type.P_yield = 0
 
     crop_type.HI_actual = 0
-    crop_type.bio_BG = 0
+    #crop_type.bio_BG = 0
```
Comment (HectorMM3) [inline RUFAS/routines/field/crop/crop.py:137]: Did we comment this out to avoid the error in carbon? I think we still need this variable as an output.

File: RUFAS/routines/field/crop/crop.py [hunk 15]
```diff
@@ -287,20 +335,27 @@ def set_grow_regimen(self, time):
                     # specified grow years have priority over cycles (specified by repeat)
                     if crop_type.repeat == 0:
                         curr_year = year - time.start_year
-                        self.grow_regimen[curr_year] = crop_type
+
+                        # if the crop is cold, set it to the latter year slot, if it is warm or
```
Comment (HectorMM3) [inline RUFAS/routines/field/crop/crop.py:339]: #replace comment with "if the crop is grown in the fall or winter then it gets placed later in the year if it is a "warm #season" -planted and grown during spring and summer or null then put it belongs in the initial year slot."

File: RUFAS/routines/field/crop/crop.py [hunk 16]
```diff
@@ -287,20 +335,27 @@ def set_grow_regimen(self, time):
...
+                            # crop slot already full
                             else:
                                 print('Cannot grow', crop_type.crop_name, 'in', str(year + curr_year) + ',',
-                                      self.grow_regimen[curr_year].crop_name, 'is already growing.')
+                                      self.grow_regimen[curr_year][planting_order].crop_name,
+                                      'is already growing.')
                             curr_year += crop_type.repeat
 
-        list(filter(lambda crop: crop.crop_name != 'null', self.grow_regimen))
+        # list(filter(lambda crop: crop.crop_name != 'null', self.grow_regimen))
```
Comment (HectorMM3) [inline RUFAS/routines/field/crop/crop.py:358]: why is it commented out?

File: RUFAS/routines/field/crop/crop_types/spring_wheat.py [hunk 17]
```diff
@@ -25,8 +26,9 @@ def __init__(self, crop_name, data):
         self.kill_year = True
         self.planted = False
         self.growing = False
+        self.killed = False
 
-        self.fix_nitrogen = False
+        self.fix_nitrogen = False  # TODO: change back
```
Comment (HectorMM3) [inline RUFAS/routines/field/crop/crop_types/spring_wheat.py:31]: please delete "# TODO: change back" it should not be changed back spring wheat does not fix nitrogen so the False value is correct.

File: RUFAS/routines/field/field_management/manure_application.py [hunk 18]
```diff
@@ -149,8 +149,8 @@ def formulate_manure_application(manure_storage, m_app):
         if desired_P != 0:
             manure_application['P_mass'] += desired_P
 
-        N_frac = 0.05 if N_frac == 0 else N_frac
-        P_frac = 0.06 if P_frac == 0 else P_frac
+        N_frac = 1 if N_frac == 0 else N_frac #TODO: this is an adjustment feature that is availabe if the manure input where total manure, not the exact values.
```
Comment (HectorMM3) [inline RUFAS/routines/field/field_management/manure_application.py:152]: add in 0.05 and 0.06 into the comments so we can have them for later.

File: RUFAS/routines/field/soil/carbon_cycling/residue_partitioning.py [hunk 19]
```diff
@@ -123,3 +141,6 @@ def residue_partitioning(soil, crop_type, weather, time):
         # S.6.B.II.9 / S.6.B.II.11
         layer.BG_struct += (AG_struct_to_BG_struct + crop_type.bio_BG * (1 - BG_met_percent)) - \
                           ((layer.BG_struct_to_C_active + layer.BG_struct_to_C_slow) + AG_struct_to_BG_struct)
+
+def ADJ_crop_type_bio_BG_fun(layer_thickness, soil_profile_depth, crop_type_bio_BG):
+    return (layer_thickness / soil_profile_depth) * crop_type_bio_BG
```
Comment (HectorMM3) [inline RUFAS/routines/field/soil/carbon_cycling/residue_partitioning.py:149]: I see a weird red symbol at the end of line 146.

**Review by HectorMM3 — CHANGES_REQUESTED** [review]
Reid, only small comments. I would proceed, and we can follow up on the details later.

File: RUFAS/routines/field/crop/crop.py (hunk 7 above)
Comment (PooyaHekmati) [reply RUFAS/routines/field/crop/crop.py:45]: In this case, we should create a git issue for it and add it here as TODO

File: RUFAS/routines/field/crop/crop.py (hunk 5 above)
Comment (PooyaHekmati) [reply RUFAS/routines/field/crop/crop.py:36]: A reviewer's comment needs to be addressed either as changes in the code or as a conversation on why the comment was wrong or not applicable. Just marking the comment resolved is not helpful.

File: RUFAS/routines/field/crop/crop.py [hunk 20]
```diff
@@ -142,9 +166,11 @@ def annual_crop_routine(crop, time):
         time: an instance of the Time class specified in classes.py
     """
 
-    # current crop is set to the next crop in the regimen
-    crop.current_crop = crop.grow_regimen[time.year - 1]
-    crop.current_crop.harvested = False
+    # current crop year is set to the next year of crops in the regimen
+    crop.current_crop_year = crop.grow_regimen[time.year - 1]
```
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop.py:170]: where is this -1 coming from?

File: RUFAS/routines/field/crop/crop.py [hunk 21]
```diff
@@ -199,12 +225,18 @@ def is_kill_year(crop, time):
             information relevant to simulating crop growth
         time: an instance of the Time class specified in classes.py
     """
-    if len(crop.grow_regimen) == time.year or \
-            crop.current_crop.crop_name != crop.grow_regimen[time.year].crop_name or \
-            crop.current_crop.crop_type == 'annual':
+
+    # the following code checks if the crop is annually grown,
```
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop.py:229]: what about moving these comments to docstrings which ended two lines above?

File: RUFAS/routines/field/crop/crop.py [hunk 22]
```diff
@@ -199,12 +225,18 @@ def is_kill_year(crop, time):
...
+    # if the crop is growing in the last year of the simulation,
+    # if the crop is not grown the next year,
+    # or if the crop is growing in a double cropping routine
+    if crop.current_crop.crop_type == 'annual' or len(crop.grow_regimen) == time.year or \
```
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop.py:233]: readability can be enhanced by moving len(crop.grow_regimen) == time.year or to next line, so each line checks one condition

File: RUFAS/routines/field/crop/crop.py [hunk 23]
```diff
@@ -243,21 +275,32 @@ def __init__(self, data, time):
...
+            # list of the crops to run during this simulation, not ordered
             self.crops_list.append(crop)
 
-        # self.alfalfa = alfalfa.Alfalfa(data)
-        # self.corn = corn.Corn(data)
-        # self.soybean = soybean.Soybean(data)
-        #
-        # self.crops_list = [self.alfalfa, self.corn, self.soybean]
-
+        # default setting
```
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop.py:293]: Please add a verb such as loading, setting, etc.

File: RUFAS/routines/field/crop/crop.py [hunk 24]
```diff
@@ -243,21 +275,32 @@ def __init__(self, data, time):
...
         self.current_crop = base_crop.BaseCrop()
+        self.current_crop_year = []
+
+        # list of the order the crops are growing in, originally set to default
+        double_cropping_limit = 2
```
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop.py:298]: avoid hardcoding by changing to 
double_cropping_limit_default_value = 2
double_cropping_limit = double_cropping_limit_default_value

File: RUFAS/routines/field/crop/crop.py (hunk 15 above)
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop.py:339]: how is this if statement being executed? where is the decision being made?

File: RUFAS/routines/field/crop/crop_types/cereal_rye.py [hunk 25]
```diff
@@ -0,0 +1,177 @@
+from .base_crop import BaseCrop
+
+
+class CerealRye(BaseCrop):
+
+    def __init__(self, crop_name, data):
+        super().__init__()
+
+        """GENERAL PLANT INFO"""
```
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop_types/cereal_rye.py:9]: Invalid format for docstrings

File: RUFAS/routines/field/crop/crop_types/cereal_rye.py [hunk 26]
```diff
@@ -0,0 +1,177 @@
...
+        self.killed = False
+
+        self.fix_nitrogen = False
+
+        # ===================================================================
+        ''' HEAT UNIT DATA '''
+
+        # input
+        self.T_base_min = 0
+        self.T_base_max = 30  # TODO
```
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop_types/cereal_rye.py:38]: Add details to TODO

File: RUFAS/routines/field/crop/crop_types/cereal_rye.py [hunk 27]
```diff
@@ -0,0 +1,177 @@
...
+        self.PHU = 996  # TODO
```
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop_types/cereal_rye.py:39]: Add details to TODO

File: RUFAS/routines/field/crop/crop_types/cereal_rye.py [hunk 28]
```diff
@@ -0,0 +1,177 @@
...
+        # ===================================================================
+        ''' LEAF AREA INDEX (LAI) DATA '''
+
+        # input
+        self.fr_PHU_1 = 0.15
+        self.fr_PHU_2 = 0.50
+        self.fr_LAI_1 = 0.01
+        self.fr_LAI_2 = 0.95
+        self.fr_PHU_sen = 0.80
+        self.fr_PHU_harvest = 1.2  # TODO
```
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop_types/cereal_rye.py:58]: Add details to TODO

File: RUFAS/routines/field/crop/crop_types/cereal_rye.py [hunk 29]
```diff
@@ -0,0 +1,177 @@
...
+
+        # output
+        self.fr_PHU = 0.0
+        self.prev_fr_PHU = 0.0
+
+        # ===================================================================
+        ''' LEAF AREA INDEX (LAI) DATA '''
+
+        # input
+        self.fr_PHU_1 = 0.15
```
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop_types/cereal_rye.py:53]: Please add reference for these values (a link to a document which explains these)

File: RUFAS/routines/field/crop/crop_types/cereal_rye.py [hunk 30]
```diff
@@ -0,0 +1,177 @@
...
+        self.fr_root = 0
+
+        # output
+        self.z_root = 0
+
+        # ===================================================================
+        ''' BIOMASS DATA '''
+
+        # input
+        self.kl = 0.65  # TODO
```
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop_types/cereal_rye.py:87]: Add details to TODO

File: RUFAS/routines/field/crop/crop_types/cereal_rye.py [hunk 31]
```diff
@@ -0,0 +1,177 @@
...
+
+        # output
+        self.biomass_actual = 0
+        self.prev_biomass_actual = 0
+
+        # ===================================================================
+        ''' Soil Water Uptake Data '''
+
+        self.beta_w = 10  # TODO: water-use distribution parameter
+        self.epco = 0.5  # TODO
```
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop_types/cereal_rye.py:104]: Add details to TODO

File: RUFAS/routines/field/crop/crop_types/cereal_rye.py [hunk 32]
```diff
@@ -0,0 +1,177 @@
...
+
+        self.beta_n = 10
+
+        self.bio_N_opt = 0
+        self.bio_N = 0
+
+        self.fr_n1 = 0.0600
+        self.fr_n2 = 0.0231
+        self.fr_n3 = 0.0130
+        self.fr_n3ish = 0.01301  # TODO
```
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop_types/cereal_rye.py:122]: Add details to TODO

File: RUFAS/routines/field/crop/crop_types/cereal_rye.py [hunk 33]
```diff
@@ -0,0 +1,177 @@
...
+
+        self.bio_P_opt = 0
+        self.bio_P = 0
+
+        self.fr_PHU_50 = 0.5
+        self.fr_PHU_100 = 1.0
+        self.fr_p1 = 0.0084
+        self.fr_p2 = 0.0032
+        self.fr_p3 = 0.0019
+        self.fr_p3ish = 0.00191  # TODO
```
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop_types/cereal_rye.py:143]: Add details to TODO

File: RUFAS/routines/field/crop/crop_types/cereal_rye.py [hunk 34]
```diff
@@ -0,0 +1,177 @@
...
+        self.HI_max = 0
+        self.HI_min = 0.2
+        self.HI_actual = 0
+        self.HI_opt = 0.40
+
+        self.harvest_eff = 0.9
+
+        self.gamma_wu = 0
+
+        self.biomass_dry_down_percent = 0.0  # TODO: Hard coded total dry down until daily method is modeled
```
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop_types/cereal_rye.py:162]: make git issue and mention it here

File: RUFAS/routines/field/crop/crop_types/cereal_rye.py [hunk 35]
```diff
@@ -0,0 +1,177 @@
...
+        self.DM_harvest_percent = 0.0001  # TODO: Hard coded dry matter percent at harvest
```
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop_types/cereal_rye.py:163]: make git issue and put it here

File: RUFAS/routines/field/crop/crop_types/soybean.py [hunk 36]
```diff
@@ -98,7 +99,7 @@ def __init__(self, crop_name, data):
         # ===================================================================
         ''' Soil Water Uptake Data '''
 
-        self.beta_w = 10  # water-use distribution parameter  # TODO: taken from corn
+        self.beta_w = 10  # water-use distribution parameter
```
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop_types/soybean.py:102]: this explanation should be inside docstrings

File: RUFAS/routines/field/crop/crop_types/triticale.py [hunk 37]
```diff
@@ -0,0 +1,177 @@
+from .base_crop import BaseCrop
+
+
+class Triticale(BaseCrop):
+
+    def __init__(self, crop_name, data):
+        super().__init__()
+
+        """GENERAL PLANT INFO"""
```
Comment (PooyaHekmati) [inline RUFAS/routines/field/crop/crop_types/triticale.py:9]: invalid format for docstring

File: RUFAS/routines/field/soil/carbon_cycling/residue_partitioning.py [hunk 38]
```diff
@@ -73,17 +77,27 @@ def residue_partitioning(soil, crop_type, weather, time):
...
 
         # below ground metabolic residue and roots
         # S.6.B.II
 
         # S.6.B.II.1
-        residue_incorp = layer.tillage_percent * soil.residue_harvest
+        if index == 0:
+            residue_incorp = layer.tillage_percent * soil.residue_harvest
+        else:
+            residue_incorp = layer.tillage_percent * 0
```
Comment (PooyaHekmati) [inline RUFAS/routines/field/soil/carbon_cycling/residue_partitioning.py]: what about residue_incorp = 0 ?

File: RUFAS/routines/field/soil/carbon_cycling/residue_partitioning.py (hunk 19 above)
Comment (PooyaHekmati) [reply RUFAS/routines/field/soil/carbon_cycling/residue_partitioning.py:149]: That shows that this is the end of the file (EOF) and no new lines are after this point.

File: RUFAS/routines/field/soil/soil.py [hunk 39]
```diff
@@ -92,7 +92,8 @@ def __init__(self, data):
         self.slope_length = data['slope_length']
         self.manning = data['manning']
         self.field_size = data['field_size']
-        self.practice_factor = data['practice_factor']
+        # self.practice_factor = data['practice_factor']
```
Comment (PooyaHekmati) [inline RUFAS/routines/field/soil/soil.py:95]: why is this commented out?

File: RUFAS/routines/field/soil/soil_erosion.py [hunk 40]
```diff
@@ -47,7 +47,11 @@ def calc_sed(soil, crop, weather, time):
     peak_runoff = calc_peak_runoff(soil, weather, time)
     K = calc_K(soil)
     C = calc_C(soil, crop)
-    P = soil.practice_factor
+
+    #P = soil.practice_factor
```
Comment (PooyaHekmati) [inline RUFAS/routines/field/soil/soil_erosion.py:51]: why is this commented out?

File: config/defintions.py [hunk 41]
```diff
@@ -0,0 +1,2 @@
+import os
+ROOT_DIR = os.path.realpath(os.path.join(os.path.dirname(__file__), '..'))
```
Comment (PooyaHekmati) [inline config/defintions.py:2]: why is this needed? where is it used?

**Review by PooyaHekmati — CHANGES_REQUESTED** [review]

**Comment by PooyaHekmati** [comment]
There are a bunch of TODOs in this PR, I commented on some of them but gave up on commenting on them all. A TODO is not desired, it is only acceptable when a git issue is created for it and it is mentioned in front of the TODO. Also, there are a bunch of docstrings with invalid formats. More importantly, this PR is heavily missing unit tests.

**Comment by KFosterReed** [comment]
I tried to run this and got a Key Error for the alfalfa 'extracted' input. It looks like you added this input variable to the alfalfa crop but it isn't added to all the crop jsons. Here is the full error report:

 

          > Initializing simulation environment from input\ARL.json
          > Traceback (most recent call last):
          >   File "C:\Users\kfr3\PycharmProjects\MASM\main.py", line 25, in <module>
          >     main()
          >   File "C:\Users\kfr3\PycharmProjects\MASM\main.py", line 20, in main
          >     simulator = SimulationEngine(input_file_path)
          >   File "C:\Users\kfr3\PycharmProjects\MASM\RUFAS\simulation_engine.py", line 23, in __init__
          >     self._initialize_simulation(input_file_path)
          >   File "C:\Users\kfr3\PycharmProjects\MASM\RUFAS\simulation_engine.py", line 153, in _initialize_simulation
          >     self.state = State(data['farm'], self.config,
          >   File "C:\Users\kfr3\PycharmProjects\MASM\RUFAS\classes.py", line 46, in __init__
          >     self.fields = Fields(data['fields'], time)
          >   File "C:\Users\kfr3\PycharmProjects\MASM\RUFAS\routines\field\field.py", line 45, in __init__
          >     self.fields[field_name] = Field(field_name, field_data, time)
          >   File "C:\Users\kfr3\PycharmProjects\MASM\RUFAS\routines\field\field.py", line 225, in __init__
          >     self.crop = Crop(crop_data, time)
          >   File "C:\Users\kfr3\PycharmProjects\MASM\RUFAS\routines\field\crop\crop.py", line 263, in __init__
          >     crop = alfalfa.Alfalfa(crop_name, crop_data)
          >   File "C:\Users\kfr3\PycharmProjects\MASM\RUFAS\routines\field\crop\crop_types\alfalfa.py", line 34, in __init__
          >     self.extracted = alfalfa_data['extracted']
          > KeyError: 'extracted'
          > 
          > Process finished with exit code 1
  
I got a similar key error for a tillage attribute of soil

File: RUFAS/routines/field/soil/soil.py (hunk 39 above)
Comment (KFosterReed) [reply RUFAS/routines/field/soil/soil.py:95]: I think this is supposed to be deleted. Is that correct?

File: RUFAS/routines/field/soil/soil_erosion.py (hunk 40 above)
Comment (KFosterReed) [reply RUFAS/routines/field/soil/soil_erosion.py:51]: Same as the other practice factor. Please delete all references

File: RUFAS/routines/field/crop/crop_types/cereal_rye.py (hunk 25 above)
Comment (KFosterReed) [reply RUFAS/routines/field/crop/crop_types/cereal_rye.py:9]: I think we should make a GIT issue to reformat the documentation in all crop types. They are all following this same style of in code documentation so let's change them all at one time and not let that hold up this pull request

File: RUFAS/routines/field/crop/crop_types/cereal_rye.py (hunk 26 above)
Comment (KFosterReed) [reply RUFAS/routines/field/crop/crop_types/cereal_rye.py:38]: Ideally they are GIT issues but let's at least have information about what the TODOs mean somewhere. As Pooya mentioned in the meeting, they can be in a google doc, written here in the comment or made into a git issue but I agree that we don't want to pull in even more cryptic TODOs than we already have

File: RUFAS/routines/field/crop/crop_types/cereal_rye.py (hunk 29 above)
Comment (KFosterReed) [reply RUFAS/routines/field/crop/crop_types/cereal_rye.py:53]: If the reference values are provided in the pseudocode documentation then values like this are OK for now

File: RUFAS/routines/field/soil/carbon_cycling/residue_partitioning.py (hunk 38 above)
Comment (KFosterReed) [reply RUFAS/routines/field/soil/carbon_cycling/residue_partitioning.py]: this makes sense to me!

File: config/defintions.py (hunk 41 above)
Comment (KFosterReed) [reply config/defintions.py:2]: Also curious...

File: RUFAS/output_handler/reports/field_report.py (hunk 2 above)
Comment (KFosterReed) [reply RUFAS/output_handler/reports/field_report.py:329]: I think we should make a GitIssue for updating the output handlers in general and leave this as is for this pull request

File: RUFAS/routines/field/crop/crop.py (hunk 7 above)
Comment (KFosterReed) [reply RUFAS/routines/field/crop/crop.py:45]: It will be good to translate these crop management functionalities into functions but I agree with Hector that the higher priority is to pull this into the master branch

File: RUFAS/routines/field/crop/crop.py (hunk 8 above)
Comment (KFosterReed) [reply RUFAS/routines/field/crop/crop.py]: I also find the logic here and below hard to follow for most of the daily_reset() function. Providing pseudocode (and possibly a diagram?) would be super helpful

File: RUFAS/routines/field/crop/crop.py (hunk 20 above)
Comment (KFosterReed) [reply RUFAS/routines/field/crop/crop.py:170]: I am also confused about the logic here. It would be helpful to have the variables used here defined in the Sphinx documentation above & in the pseudocode (or at least one or the other).

File: RUFAS/routines/field/crop/crop.py (hunk 21 above)
Comment (KFosterReed) [reply RUFAS/routines/field/crop/crop.py:229]: There are several places like this where the code can be improved for readability. I think tagging them with a TODO and linking them into a single GITissue would be great

**Review by KFosterReed — CHANGES_REQUESTED** [review]
This seems very close to me but there are a few things that need to be done before we can merge it in: 
1) The logic of how the crop rotations are managed needs to be more clear. That is the heart of the new functionality and it is very difficult to follow. Whether it is with external documentation, updates to the in code documentation, or changes to the code itself does not matter to me. 
2) Please create git issues that group opportunities for improving the code clarity and 'cleanness' and add TODOs where identified
3) The model needs to run on all .json files

**Comment by KFosterReed** [comment]
@rhensen I am guessing the new PR replaces this one? If so, can we close this one?
